### ISSUE TITLE
Reverse mapping: tabular MDP state to game state

### ISSUE THREAD

**Issue by JosselinSomervilleRoberts** [cassidylaidlaw/effective-horizon#2]
For another project, we would like to visualize our Q-value function using the tabular MDP (MiniGrid envs) but we do not know how to gor from an MDP state to a game state. Is there a way to do this? Is this even doable?
Thanks!

**Comment by cassidylaidlaw** [comment]
Hi Josselin, thanks for the interest in the project! Here is some code that will find an environment state in the gym env from a tabular state:

```
import numpy as np
import gymnasium as gym
import matplotlib.pyplot as plt
from effective_horizon.mdp_utils import load_mdp, get_sparse_mdp, run_value_iteration

mdp_name = "MiniGrid-Empty-8x8-v0"  # Replace with the name of the BRIDGE MDP.
tabular_state = 63  # Replace with the index of the tabular state.

transitions, rewards = load_mdp("data/bridge_dataset/mdps/MiniGrid-Empty-8x8-v0/consolidated.npz")
num_states, num_actions = transitions.shape
done_state = num_states - 1

one_hot_rewards = np.zeros((num_states, num_actions))
one_hot_rewards[tabular_state, :] = 1

sparse_transitions, rewards_vector = get_sparse_mdp(transitions, one_hot_rewards)
vi = run_value_iteration(sparse_transitions, rewards_vector, horizon=100)

env = gym.make(f"BRIDGE/{mdp_name}")
env.reset()

state, t = 0, 0
while state != tabular_state:
    action = np.argmax(vi.optimal_qs[t, state, :])
    env.step(action)
    state = transitions[state, action]
    t += 1

env.unwrapped.render_mode = "rgb_array"
plt.imshow(env.render())
```

The idea is that if you can find a sequence of actions that leads to the tabular state in the tabular representation, then you can take the same sequence of actions in the gym env and it will get to the same state. Here that's accomplished by setting a special reward function which gives 1 reward only for taking an action in the specified state. Then, running value iteration on this reward function will find give a Q-function which can be used to find a sequence of actions leading to the desired state.

Once the gym env is in the desired state, you can render it, look at its internal state, etc. Hope this helps!

**Comment by JosselinSomervilleRoberts** [comment]
Thank you so much, this is very useful!
I still have a few questions.
- When I run the script on `MiniGrid-Empty-8x8-v0` it seems to work well but on `MiniGrid-DoorKey-8x8-v0` the done state does not seem to be `num_states - 1` (See picture below)
![image](https://github.com/cassidylaidlaw/effective-horizon/assets/73412738/c8b8ef20-2f24-4f0e-a0df-415ce430bb5d)
- This method is good for visualizing one specific state in an environment that do not have many states. What I really need is to have a mapping of all the tabular states to all the in-game states. I guess I could just rollout a policy and save it every time but it would be really expensive. Is there not something simpler?

Thanks a lot for your help and fast response!

**Comment by cassidylaidlaw** [comment]
Thanks for finding that issue with the done state not matching up! It looks like there was actually a bug in the code for generating the tabular representations for the MiniGrid MDPs. It was fixed a while ago but for some reason I never regenerated the tabular representations for the non-shaped MiniGrid envs.

I am now regenerating those and will send you the fixed versions when they're ready. I'm also including the screens and state representations in the tabular MDP file so it will be easier to quickly get the screen for an arbitrary tabular state.

**Comment by JosselinSomervilleRoberts** [comment]
This is amazing thank you very much!
In the meantime, I created a small script that gives a simple visualization by constructing a simplified mapping between the tabular state and the game state (this is a simplification). The code is available here for anyone interested: https://github.com/JosselinSomervilleRoberts/TabularRL-Experiments

You can then use the mapping to visualize the V-values and optimal path for example:
![image](https://github.com/cassidylaidlaw/effective-horizon/assets/73412738/88f1205c-dcb5-42b2-8ca7-591cf3ccaead)
Hope this helps some people and thank you so much for your help, having this new dataset will be amazing!

**Comment by JosselinSomervilleRoberts** [comment]
Hi @cassidylaidlaw , any update on this new dataset?

**Comment by cassidylaidlaw** [comment]
Hi Josselin, we're still working on getting the next version of the dataset together, but in the mean time this zip file contains the new versions of the MiniGrid MDPs at least: https://drive.google.com/file/d/1GER40Qqf-RKugdP-Y9XbFkJH22Hs_gz3/view?usp=sharing

You can use the newer versions of the MDP representations as follows:

```python3
mdp_name = "MiniGrid-DoorKey-5x5-v0"  # Replace with the name of the BRIDGE MDP.
tabular_state = 25  # Replace with the index of the tabular state.
mdp_fname = f"mdps/{mdp_name}/consolidated.npz"
mdp = np.load(mdp_fname)

# Get the rendered image for the tabular state.
plt.imshow(mdp["screens"][mdp["screen_mapping"][tabular_state]]) #.reshape(3, 256, 256).transpose((2, 1, 0)))
plt.show()

# Set the environment to the tabular state.
env = gym.make(f"BRIDGE/{mdp_name}")
env.reset()
state_bytes = mdp["states"][tabular_state, :mdp["state_lengths"][tabular_state]].tobytes()
env.set_state(state_bytes)
plt.imshow(env.render())
plt.show()
```

Let me know if you have any questions or run into any issues!

**Comment by JosselinSomervilleRoberts** [comment]
That's amazing thanks!

**Comment by JosselinSomervilleRoberts** [comment]
Quick question? how did you register the "BRIDGE" namespace for `gym`?

**Comment by cassidylaidlaw** [comment]
All the environments should be registered from this file: https://github.com/cassidylaidlaw/effective-horizon/blob/main/effective_horizon/envs/deterministic_registration.py

You might have to add `import effective_horizon` to make sure they are registered.

**Comment by cassidylaidlaw** [comment]
Just as an FYI, a new version of the dataset has been released (v1.1) with the screens and binary states for more of the MDPs: https://zenodo.org/records/10966777

There's also a getting started notebook now that shows how to use the tabular MDPs: https://colab.research.google.com/github/cassidylaidlaw/effective-horizon/blob/main/getting_started.ipynb
